Makefile

```diff
@@ -32,18 +32,43 @@ SUB := $(addsuffix .sub, $(ALL))
 build: build.sub
 build-linux: build-linux.sub
 build-darwin: build-darwin.sub
-clean: clean.sub
 docker: docker.sub
 lint: lint.sub
 push: push.sub
 test: test.sub
 
+clean: clean.sub
+	rm -f .dockerInit
+	rm -f .buildImage
+	docker rmi -f buildImage > /dev/null 2>&1 || true
+
+# Use this target when you want to build everything using docker containers.
+# Good for cases when you don't have the tools installed (like glide or go).
+docker-all: .buildImage
+	docker run --rm -ti \
+	  -v $(PWD):/go/src/github.com/kubernetes-incubator/service-catalog \
+	  buildimage \
+	  make .dockerInit all
+
+# .dockerInit tells us if our vendor stuff if out of date or not.
+# And if so we'll run init under our docker build.  For non-docker builds
+# it is assumed you'll run "make init" manually.
+.dockerInit: glide.yaml
+	make init
+	echo > .dockerInit
+
+# .buildImage is used to know when the docker image ("buildimage") is out
+# of date with the Dockerfile.
+.buildImage: hack/Dockerfile
+	docker build -t buildimage - < hack/Dockerfile
+	echo > .buildImage
```

Comment:
Phony targets run all the time. @duglin is trying to avoid having to rebuild the image on each build. The .buildImage file is used to represent "I have Docker image built at this time" so if Dockerfile gets updated, make will do the right thing.

`.PHONY` would work but lead to rebuilds of the build image on each `make docker-all`